### PR TITLE
SW-1805-change-pico-bootloader-code

### PR DESCRIPTION
**Pull request by amrithvenkat**
Updates baud rate, starts flash process when data available on UART line

### CODE REVIEW

File: main.c [hunk 1]
```diff
@@ -686,13 +686,30 @@ int main(void)
...
+	struct image_header *hdr = (struct image_header *)(XIP_BASE + IMAGE_HEADER_OFFSET);
+
+	gpio_set_function(UART_TX_PIN, GPIO_FUNC_UART);
+	gpio_set_function(UART_RX_PIN, GPIO_FUNC_UART);
+	uart_init(uart0, UART_BAUD);
+	uart_set_hw_flow(uart0, false, false);
 
 	sleep_ms(10);
+	
+    while (uart_is_readable(uart0)){uart_getc(uart0);}; //clear Buffer
```
Comment (ahmed-mrbeam) [inline main.c]: The alignment of the code doesn't seem consistent

File: main.c [hunk 2]
```diff
@@ -702,16 +719,17 @@ int main(void)
 	}
 
 	DBG_PRINTF_INIT();
-
-	uart_init(uart0, UART_BAUD);
-	gpio_set_function(UART_TX_PIN, GPIO_FUNC_UART);
-	gpio_set_function(UART_RX_PIN, GPIO_FUNC_UART);
-	uart_set_hw_flow(uart0, false, false);
+	
+	//uart_init(uart0, UART_BAUD);
```
Comment (ahmed-mrbeam) [inline main.c]: why commenting this part, why not remove ?

File: main.c [hunk 3]
```diff
@@ -702,16 +719,17 @@ int main(void)
...
+	//gpio_set_function(UART_TX_PIN, GPIO_FUNC_UART);
+	//gpio_set_function(UART_RX_PIN, GPIO_FUNC_UART);
+	//uart_set_hw_flow(uart0, false, false);
 
 	struct cmd_context ctx;
 	uint8_t uart_buf[(sizeof(uint32_t) * (1 + MAX_NARG)) + MAX_DATA_LEN];
 	ctx.uart_buf = uart_buf;
 	enum state state = STATE_WAIT_FOR_SYNC;
+	watchdog_reboot(0, 0, 4000); //Set WD to 4s
```
Comment (ahmed-mrbeam) [inline main.c]: 4000 is a magic number, you should use 
`#define WD_TIMEOUT_MS (4000)`

File: main.c [hunk 4]
```diff
@@ -686,13 +686,30 @@ int main(void)
...
+	gpio_set_function(UART_RX_PIN, GPIO_FUNC_UART);
+	uart_init(uart0, UART_BAUD);
+	uart_set_hw_flow(uart0, false, false);
 
 	sleep_ms(10);
+	
+    while (uart_is_readable(uart0)){uart_getc(uart0);}; //clear Buffer
+    
+	//Wait loop
+    bool no_rec = true;
```
Comment (ahmed-mrbeam) [inline main.c]: Variable name is ambiguous

File: main.c [hunk 5]
```diff
@@ -686,13 +686,30 @@ int main(void)
...
 
-	gpio_init(BOOTLOADER_ENTRY_PIN);
-	gpio_pull_up(BOOTLOADER_ENTRY_PIN);
-	gpio_set_dir(BOOTLOADER_ENTRY_PIN, 0);
+	struct image_header *hdr = (struct image_header *)(XIP_BASE + IMAGE_HEADER_OFFSET);
+
+	gpio_set_function(UART_TX_PIN, GPIO_FUNC_UART);
+	gpio_set_function(UART_RX_PIN, GPIO_FUNC_UART);
+	uart_init(uart0, UART_BAUD);
+	uart_set_hw_flow(uart0, false, false);
```
Comment (ahmed-mrbeam) [inline main.c]: you shouldn't use uart0, you should use a define instead

File: main.c [hunk 6]
```diff
@@ -686,13 +686,30 @@ int main(void)
...
 	sleep_ms(10);
+	
+    while (uart_is_readable(uart0)){uart_getc(uart0);}; //clear Buffer
+    
+	//Wait loop
+    bool no_rec = true;
+
+    uart_puts(uart0, "Boot\r\n");  //Header
+
+	for(int i = 0; i < 10000; i++){
```
Comment (ahmed-mrbeam) [inline main.c]: 10000 Magic number , we need a define for that

File: main.c [hunk 7]
```diff
@@ -686,13 +686,30 @@ int main(void)
...
+       busy_wait_us(100);
+       if (uart_is_readable(uart0)){     
+           no_rec = false;
+		   i = 11000;
```
Comment (ahmed-mrbeam) [inline main.c]: we need to know what is the purpose of this umber

File: main.c [hunk 8]
```diff
@@ -686,13 +686,30 @@ int main(void)
...
+    
+	//Wait loop
+    bool no_rec = true;
+
+    uart_puts(uart0, "Boot\r\n");  //Header
+
+	for(int i = 0; i < 10000; i++){
+       busy_wait_us(100);
+       if (uart_is_readable(uart0)){     
+           no_rec = false;
```
Comment (ahmed-mrbeam) [inline main.c]: This is not used anywhere afterwards, what is the purpose of it ?

File: main.c [hunk 9]
```diff
@@ -29,14 +29,15 @@
...
 // The bootloader can be entered in three ways:
 //  - BOOTLOADER_ENTRY_PIN is low
-//  - Watchdog scratch[5] == BOOTLOADER_ENTRY_MAGIC && scratch[6] == ~BOOTLOADER_ENTRY_MAGIC
+//  - Watchdog scratch[5] == BOOTLOADER_ENTRY_PIN && scratch[6] == ~BOOTLOADER_ENTRY_MAGIC
 //  - No valid image header
-#define BOOTLOADER_ENTRY_PIN 15
+// #define BOOTLOADER_ENTRY_PIN 15
 #define BOOTLOADER_ENTRY_MAGIC 0xb105f00d
 
+#define UART_ID 	(uart0)
```
Comment (ahmed-mrbeam) [inline main.c]: why ID  ? why not CH for Channel ?

Comment (amrithvenkat) [reply main.c]: Refered an autogenerated sample from Pico-project generator tool. it used ID and i followed that convention

File: main.c (hunk 8 above)
Comment (amrithvenkat) [reply main.c]: reflected the changes in the latest commit. it is used

File: main.c (hunk 7 above)
Comment (amrithvenkat) [reply main.c]: Changed it to use a break statement. Basti intended to use it to break out of the loop.

File: main.c (hunk 6 above)
Comment (amrithvenkat) [reply main.c]: The purpose of the loop is to run for 1 second split into 10000 times of each 100 uS per run. what should we name it?

File: main.c (hunk 5 above)
Comment (amrithvenkat) [reply main.c]: definition replaced in the code

File: main.c (hunk 3 above)
Comment (amrithvenkat) [reply main.c]: This watchdog reboot was a coding mistake and had it removed in the latest commit

File: main.c (hunk 4 above)
Comment (amrithvenkat) [reply main.c]: it stands for Nothing received. the boolean is set to false if something is received in the uart line while in the bootloader. it is then used in the if statement that follows

Comment (ahmed-mrbeam) [reply main.c]: no_trigger_received

File: main.c [hunk 10]
```diff
@@ -29,14 +29,15 @@
 
 // The bootloader can be entered in three ways:
 //  - BOOTLOADER_ENTRY_PIN is low
-//  - Watchdog scratch[5] == BOOTLOADER_ENTRY_MAGIC && scratch[6] == ~BOOTLOADER_ENTRY_MAGIC
+//  - Watchdog scratch[5] == BOOTLOADER_ENTRY_PIN && scratch[6] == ~BOOTLOADER_ENTRY_MAGIC
```
Comment (ahmed-mrbeam) [inline main.c]: This `BOOTLOADER_ENTRY_PIN` is changed here but `BOOTLOADER_ENTRY_MAGIC` is still used in function `should_stay_in_bootloader`

File: main.c [hunk 11]
```diff
@@ -29,14 +29,15 @@
...
 //  - No valid image header
-#define BOOTLOADER_ENTRY_PIN 15
+// #define BOOTLOADER_ENTRY_PIN 15
 #define BOOTLOADER_ENTRY_MAGIC 0xb105f00d
```
Comment (ahmed-mrbeam) [inline main.c:36]: what is actually meant by entry magic ?

File: main.c [hunk 12]
```diff
@@ -29,14 +29,15 @@
 
 // The bootloader can be entered in three ways:
 //  - BOOTLOADER_ENTRY_PIN is low
-//  - Watchdog scratch[5] == BOOTLOADER_ENTRY_MAGIC && scratch[6] == ~BOOTLOADER_ENTRY_MAGIC
+//  - Watchdog scratch[5] == BOOTLOADER_ENTRY_PIN && scratch[6] == ~BOOTLOADER_ENTRY_MAGIC
 //  - No valid image header
-#define BOOTLOADER_ENTRY_PIN 15
+// #define BOOTLOADER_ENTRY_PIN 15
```
Comment (ahmed-mrbeam) [inline main.c:35]: what is actually meant by entry magic ??

File: main.c [hunk 13]
```diff
@@ -749,4 +765,4 @@ int main(void)
 	}
 
 	return 0;
-}
+}
```
Comment (ahmed-mrbeam) [inline main.c]: new line missing at the end of file

File: .gitignore [hunk 14]
```diff
@@ -0,0 +1,2 @@
+build
+.vscode
```
Comment (ahmed-mrbeam) [inline .gitignore]: new line missing at the end of file

File: main.c [hunk 15]
```diff
@@ -686,15 +687,35 @@ int main(void)
...
-	sleep_ms(10);
-
 	struct image_header *hdr = (struct image_header *)(XIP_BASE + IMAGE_HEADER_OFFSET);
+    
+    gpio_set_function(UART_TX_PIN, GPIO_FUNC_UART);
+	gpio_set_function(UART_RX_PIN, GPIO_FUNC_UART);
+	uart_init(UART_ID, UART_BAUD);
+	uart_set_hw_flow(UART_ID, false, false);
+    
+	sleep_ms(10);  // Bring UP Serial Lines
```
Comment (ahmed-mrbeam) [inline main.c]: The comment should be:
`// waiting for Serial lines to fully intialize`

File: main.c [hunk 16]
```diff
@@ -686,15 +687,35 @@ int main(void)
...
+ 
+    while (uart_is_readable(UART_ID))
+		{
+		uart_getc(UART_ID);
+		}; //clear Buffer
```
Comment (ahmed-mrbeam) [inline main.c]: what is is the comment referring to ?, which buffer ?

File: readme.txt [hunk 17]
```diff
@@ -0,0 +1,37 @@
+POINTS TO NOTE:
+
+* bootloader.ld
+
+-> FLASH(rx) : ORIGIN = 0x10000000, LENGTH = 12k //maintain the length as multiple of 4k
+-> Make similar changes in the other .ld files and CMakeLists.txt file
+    combined.ld
+    standalone.ld
+
+Refer the branch - https://github.com/usedbytes/rp2040-serial-bootloader/tree/bl32k and the changes.
```
Comment (ahmed-mrbeam) [inline readme.txt:10]: is this correct ?

Comment (amrithvenkat) [reply readme.txt:10]: Yes. If we need to use a bootloader which occupies more space i.e > 12k assigned for it during compile time, then the implementation in that branch can be used/referred and changes implemented to suit the needs. 

This issue arises when the Bootloader C code is compiled without optimization. Cmake has the optimisation set to "Debug" which has less optimization. Ideally should use the "Release" which will make the size of the code < or equal to 12k

File: main.c [hunk 18]
```diff
@@ -686,15 +687,35 @@ int main(void)
...
+    gpio_set_function(UART_TX_PIN, GPIO_FUNC_UART);
+	gpio_set_function(UART_RX_PIN, GPIO_FUNC_UART);
+	uart_init(UART_ID, UART_BAUD);
+	uart_set_hw_flow(UART_ID, false, false);
+    
+	sleep_ms(10);  // Bring UP Serial Lines
+ 
+    while (uart_is_readable(UART_ID))
+		{
+		uart_getc(UART_ID);
```
Comment (ahmed-mrbeam) [inline main.c]: check if this is needed, I assume it corrupts the first sync attempt

File: main.c (hunk 4 above)
Comment (ahmed-mrbeam) [reply main.c]: `nothing_received`

File: bl_config.h [hunk 19]
```diff
@@ -0,0 +1,13 @@
+#ifndef BL_CONFIG_H_INCLUDED
+#define BL_CONFIG_H_INCLUDED
+
+#include "bl_version.h"
+
+#define BL_WD_TO_MAX_MS         8000
+
+#define STR_INDIR(x) #x
```
Comment (ahmed-mrbeam) [inline bl_config.h]: Please create a new header file called bl_utils.h and move those two macros to it.
Also add a comment there explaining the use case of them

File: main.c [hunk 20]
```diff
@@ -687,30 +689,23 @@ int main(void)
...
+    bool nothing_received = true;
+    uart_puts(UART_ID, BL_IDENTIFIER_STR);  //Header
 
-	for(int i = 0; i < 10000; i++){
-       busy_wait_us(100);
-       if (uart_is_readable(uart0)){     
-           no_rec = false;
-		   i = 11000;
-		}
+	if(uart_is_readable_within_us(UART_ID,1000000)) {
```
Comment (ahmed-mrbeam) [inline main.c]: Magic number, kindly create a define for that in the config header file

**Review by ahmed-mrbeam — CHANGES_REQUESTED** [review]

File: bl_version.h [hunk 21]
```diff
@@ -0,0 +1,9 @@
+#ifndef BL_VERSION_H_INCLUDED
+#define BL_VERSION_H_INCLUDED
+
+#define VERSION_MAJOR       "00"
+#define VERSION_MINOR       "01"
+#define VERSION_PATCH       "00"
+#define VERSION_OTHERS      "feature"
```
Comment (ahmed-mrbeam) [inline bl_version.h]: Please leave the others string as empty string for now `""`

File: main.c (hunk 20 above)
Comment (amrithvenkat) [reply main.c]: done

File: bl_version.h (hunk 21 above)
Comment (amrithvenkat) [reply bl_version.h]: done

File: bl_utils.h [hunk 22]
```diff
@@ -0,0 +1,12 @@
+#ifndef BL_UTILS_H_INCLUDED
+#define BL_UTILS_H_INCLUDED
+
+#define BL_WD_TO_MAX_MS         (8000)
```
Comment (ahmed-mrbeam) [inline bl_utils.h]: this should be in config header

File: bl_utils.h [hunk 23]
```diff
@@ -0,0 +1,12 @@
...
+#define UART_READABLE_MAX_US    (1000000)
```
Comment (ahmed-mrbeam) [inline bl_utils.h]: this should be in config header, and also we should name it to something related to the context, here we wait for stay in BL trigger so :
`BL_STAY_IN_BL_UART_TRIGGER_TO_MAX_US`

File: main.c [hunk 24]
```diff
@@ -31,12 +32,13 @@
...
 //  - No valid image header
-#define BOOTLOADER_ENTRY_PIN 15
+// #define BOOTLOADER_ENTRY_PIN 15
 #define BOOTLOADER_ENTRY_MAGIC 0xb105f00d
 
+#define UART_ID 	(uart0)
 #define UART_TX_PIN 0
 #define UART_RX_PIN 1
-#define UART_BAUD   115200
+#define UART_BAUD   (115200)
```
Comment (ahmed-mrbeam) [inline main.c]: this should move to config as well

File: bl_utils.h (hunk 23 above)
Comment (amrithvenkat) [reply bl_utils.h]: UART_WAIT_FOR_TRIGGER_TOUT_US

**Review by ahmed-mrbeam — APPROVED** [review]

**Pull request merged**